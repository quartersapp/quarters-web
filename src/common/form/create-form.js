import React, { Component } from 'react'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'
import PropTypes from 'prop-types'
import { registerForm, deregisterForm } from './actions'
import { createFormValuesSelector } from './selectors'

export const createForm = name => {
  class Form extends Component {
    componentWillMount () {
      this.props.registerForm(name)
    }

    componentWillUnmount () {
      this.props.deregisterForm(name)
    }

    /**
     * Set the form name to the context so that field
     * components have access to the form name
     */
    getChildContext () {
      return { form: name }
    }

    handleSubmit = e => {
      e.preventDefault()
      this.props.onSubmit && this.props.onSubmit(this.props.values)
    }

    render () {
      return (
        <form onSubmit={this.handleSubmit}>
          {this.props.children}
        </form>
      )
    }
  }

  Form.childContextTypes = {
    form: PropTypes.string
  }

  Form.propTypes = {
    onSubmit: PropTypes.func,
    children: PropTypes.node
  }

  return Form
}

export default name => connect(
  createStructuredSelector({ values: createFormValuesSelector(name) }),
  { registerForm, deregisterForm }
)(createForm(name))
