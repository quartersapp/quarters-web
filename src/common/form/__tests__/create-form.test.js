/* eslint-env jest */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { mount } from 'enzyme'
import renderer from 'react-test-renderer'
import Immutable from 'seamless-immutable'
import { createForm } from '../create-form'

const noop = () => {}
const FORM_NAME = 'testForm'

describe('createForm', () => {
  const Form = createForm(FORM_NAME)

  it('matches snapshot', () => {
    const tree = renderer.create(
      <Form registerForm={noop}>
        Form goes here
      </Form>
    )
    expect(tree.toJSON()).toMatchSnapshot()
  })

  it('sets the form name to the context', () => {
    /* Couldn't get context testing with enzyme working, so this will have to do */
    class ChildComponent extends Component {
      render () {
        return null
      }
    }

    ChildComponent.contextTypes = {
      form: PropTypes.string
    }

    const wrapper = mount(
      <Form registerForm={noop}>
        <ChildComponent />
      </Form>
    )

    const child = wrapper.find(ChildComponent).get(0)
    expect(child.context.form).toEqual(FORM_NAME)
  })

  it('dispatches register and deregister actions on mount/unmount', () => {
    const registerForm = jest.fn()
    const deregisterForm = jest.fn()

    const wrapper = mount(
      <Form registerForm={registerForm} deregisterForm={deregisterForm} />
    )
    expect(registerForm).toBeCalled()
    expect(deregisterForm).not.toBeCalled()

    wrapper.unmount()
    expect(deregisterForm).toBeCalled()
  })

  it('calls props.handleSubmit with the form values onSubmit', () => {
    const handleSubmit = jest.fn()
    const preventDefault = jest.fn()
    const formValues = {
      email: 'test@example.com',
      password: '12345'
    }

    const wrapper = mount(
      <Form registerForm={noop} onSubmit={handleSubmit} values={formValues} />
    )

    wrapper.find('form').simulate('submit', { preventDefault })
    expect(handleSubmit).toBeCalledWith(formValues)
    expect(preventDefault).toBeCalled()
  })
})
