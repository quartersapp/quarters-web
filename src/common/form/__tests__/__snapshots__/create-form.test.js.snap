// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createForm matches snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  Form goes here
</form>
`;
